File: BigOBlue/Lecture8_Dijkstras_Algrorithm/Commandos.cpp
#include <iostream>
#include <vector>
#include <queue>
const int INF = 1e9 + 7;
using namespace std;

vector<pair<int, int>> graph[105];

vector<int> Dijkstra(int s, int n) {
    vector<int> dist(n, INF);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, s});
    dist[s] = 0;

    while(!pq.empty()) {
        int u = pq.top().second;
        int w = pq.top().first;
        pq.pop();

        if(w > dist[u]) continue;

        for(auto &v : graph[u]) {
            if(w + v.first < dist[v.second]) {
                dist[v.second] = w + v.first;
                pq.push({dist[v.second], v.second});
            }
        }
    }
    return dist;
}

int main() {
    int t, n, r, tc = 1;
    cin >> t;
    while(t--) {
        cin >> n >> r;

        for(int i = 0; i < 105; i++) {
            graph[i].clear();
        }

        for(int i = 0; i < r; i++) {
            int u, v;
            cin >> u >> v;
            graph[u].push_back({1, v});
            graph[v].push_back({1, u});
        }

        int s, d;
        cin >> s >> d;
        vector<int> distS = Dijkstra(s, n);
        vector<int> distD = Dijkstra(d, n);

        int res = 0;
        for(int i = 0; i < n; i++) {
            res = max(res, distS[i] + distD[i]);
        }
        cout << "Case " << tc++ << ": " << res << endl;
    }
}

File: BigOBlue(Python)/Dijkstra/Commandos.py
import queue
graph = [[]]
dist = []
INF = 10 ** 9

class Edge:
    def __init__(self, weight, id):
        self.weight = weight
        self.id = id
    def __lt__(self, other):
        return self.weight < other.weight
    
def Dijkstra(s, N):
    dist = [INF] * (N + 1)
    pq = queue.PriorityQueue()
    pq.put(Edge(0, s))
    dist[s] = 0
    
    while not pq.empty():
        top = pq.get()
        u = top.id
        w = top.weight
        if w > dist[u]:
            continue
    
        for v in graph[u]:
            if w + v.weight < dist[v.id]:
                dist[v.id] = w + v.weight
                pq.put(Edge(dist[v.id], v.id))
    return dist

T = int(input())
for tc in range(1, T + 1):
    N = int(input())
    R = int(input())
    graph = [[] for _ in range(N + 1)]
    for _ in range(R):
        u, v = map(int, input().split())
        graph[u].append(Edge(1, v))
        graph[v].append(Edge(1, u))
        
    s, d = map(int, input().split())
    distS = Dijkstra(s, N)
    distD = Dijkstra(d, N)
    res = 0
    for i in range(N):
        res = max(res, distS[i] + distD[i])
    print("Case {}: {}".format(tc, res))
